clang/tools/clang-sycl-linker/ClangSYCLLinker.cpp

```diff
@@ -264,44 +271,86 @@ Expected<SmallVector<std::string>> getSYCLDeviceLibs(const ArgList &Args) {
   return DeviceLibFiles;
 }
 
-/// Link all device library files and input file into one LLVM IR file. This
-/// linking is performed using llvm-link tool.
-/// 'InputFiles' is the list of all LLVM IR device input files.
-/// 'Args' encompasses all arguments required for linking device code and will
-/// be parsed to generate options required to be passed into llvm-link tool.
-static Expected<StringRef> linkDeviceLibFiles(StringRef InputFile,
-                                              const ArgList &Args) {
-  llvm::TimeTraceScope TimeScope("LinkDeviceLibraryFiles");
-
-  auto SYCLDeviceLibFiles = getSYCLDeviceLibs(Args);
-  if (!SYCLDeviceLibFiles)
-    return SYCLDeviceLibFiles.takeError();
-  if ((*SYCLDeviceLibFiles).empty())
-    return InputFile;
-
-  Expected<std::string> LLVMLinkPath =
-      findProgram(Args, "llvm-link", {getMainExecutable("llvm-link")});
-  if (!LLVMLinkPath)
-    return LLVMLinkPath.takeError();
-
+/// Following tasks are performed:
+/// 1. Link all SYCL device bitcode images into one image. Device linking is
+/// performed using the linkModules API.
+/// 2. Gather all SYCL device library bitcode images.
+/// 3. Link all the images gathered in Step 2 with the output of Step 1 using
+/// linkModules API. LinkOnlyNeeded flag is used.
+Expected<StringRef> linkDeviceCode(ArrayRef<OffloadFile> BitcodeInputFiles,
+                                   const ArgList &Args, LLVMContext &C) {
+  llvm::TimeTraceScope TimeScope("SYCL Link Device Code");
+  // Gather SYCL device library files.
+  auto SYCLDeviceLibFilenames = getSYCLDeviceLibs(Args);
+  if (!SYCLDeviceLibFilenames)
+    return SYCLDeviceLibFilenames.takeError();
   // Create a new file to write the linked device file to.
-  auto OutFileOrErr =
+  auto BitcodeOutput =
       createTempFile(Args, sys::path::filename(OutputFile), "bc");
-  if (!OutFileOrErr)
-    return OutFileOrErr.takeError();
-
-  SmallVector<StringRef, 8> CmdArgs;
-  CmdArgs.push_back(*LLVMLinkPath);
-  CmdArgs.push_back("-only-needed");
-  CmdArgs.push_back(InputFile);
-  for (auto &File : *SYCLDeviceLibFiles)
-    CmdArgs.push_back(File);
-  CmdArgs.push_back("-o");
-  CmdArgs.push_back(*OutFileOrErr);
-  CmdArgs.push_back("--suppress-warnings");
-  if (Error Err = executeCommands(*LLVMLinkPath, CmdArgs))
-    return std::move(Err);
-  return *OutFileOrErr;
+  if (!BitcodeOutput)
+    return BitcodeOutput.takeError();
+  int FD = -1;
+  if (std::error_code EC = sys::fs::openFileForWrite(*BitcodeOutput, FD))
+    return errorCodeToError(EC);
+  if (DryRun || Verbose) {
+    std::for_each((*SYCLDeviceLibFilenames).begin(),
+                  (*SYCLDeviceLibFilenames).end(),
+                  [](std::string &Filename) { Filename += "; "; });
+    std::string LibInputs =
+        std::accumulate((*SYCLDeviceLibFilenames).begin(),
+                        (*SYCLDeviceLibFilenames).end(), std::string(""));
```

Comment:
It can be done with a single std::accumulate call with something like:
```suggestion
    std::string LibInputs =
        std::accumulate(std::next((*SYCLDeviceLibFilenames).begin()),
                        (*SYCLDeviceLibFilenames).end(), (*SYCLDeviceLibFilenames).front(),
                        [](string a, string b) {
                          return std::move(a) + "; " + std::move(b);
                        });
```
Basically, use the overload with binary operator.

BTW, why not use `,` as files delimiter?

Reply:
That's a great suggestion. Thanks much. Will apply this.
Reason for using ';'
This seemed more readable: a;b;c;d; than a,b,c,d
I think, with your change, it might just be a,b,c,d

Will check that out.

Thanks